typing/uniquenessAnalysis.ml

```diff
@@ -0,0 +1,557 @@
+      | OneParent(p) ->
+        let p, uenv = add_anon_mproj p mproj uenv in
+        List.fold_left2
+          (fun uenv pat' i ->
+             pat_to_map_ pat' (OneParent p)
+              (Some (Projection.Construct_field(Longident.last lbl.txt, i)))
+              uenv)
+          uenv ps (List.init (List.length ps) Fun.id)
+      | TupleParent _ -> assert false
+    end
+  | Tpat_variant(lbl, mpat, _) -> begin
+      match mpat with
+      | Some pat' -> let parent, uenv = match mproj, parent with
+        | Some _, OneParent p ->
+          let p, uenv = add_anon_mproj p mproj uenv in OneParent(p), uenv
+        | _, _ -> parent, uenv in
+        pat_to_map_ pat' parent (Some (Projection.Variant_field lbl)) uenv
+      | None -> uenv
+      end
+  | Tpat_record(ps, _) -> begin
+      match ps with | [] -> uenv | _ ->
+      match parent with
+      | NoParent ->
+        List.fold_left (fun uenv (_, _, pat') -> pat_to_map_ pat' parent None uenv) uenv ps
+      | OneParent(p) ->
+        let p, uenv = add_anon_mproj p mproj uenv in
+        List.fold_left
+          (fun uenv (_ , l, pat') ->
+             pat_to_map_ pat' (OneParent p)
+               (Some (Projection.Record_field l.lbl_name)) uenv)
+          uenv ps
+      | TupleParent _ -> assert false
+    end
+  | Tpat_array(ps) -> begin
+      match ps with | [] -> uenv | _ ->
+      match parent with
+      | NoParent ->
+        List.fold_left (fun uenv pat' -> pat_to_map_ pat' parent None uenv) uenv ps
+      | OneParent(p) ->
+        let p, uenv = add_anon_mproj p mproj uenv in
+        List.fold_left2
+          (fun uenv pat' i ->
+              pat_to_map_ pat' (OneParent p) (Some (Projection.Array_field i)) uenv)
+          uenv ps (List.init (List.length ps) Fun.id)
+      | TupleParent _ -> assert false
+    end
+  | Tpat_lazy(pat') -> pat_to_map_ pat' parent mproj uenv
+  | Tpat_or(a, b, _) ->
+    pat_to_map_ a parent mproj (pat_to_map_ b parent mproj uenv)
+
+let pat_to_map (pat : Typedtree.pattern) parent uenv =
+  pat_to_map_ pat parent None uenv
+
+(* We ignore exceptions in uniqueness analysis. *)
+let comp_pat_to_map
+      (pat : Typedtree.computation Typedtree.general_pattern) parent uenv =
+  match split_pattern pat with
+  | Some pat', _ -> pat_to_map pat' parent uenv
+  | None, _ -> uenv
+
+let ident_option_from_path p =
+  match p with
+  | Path.Pident id -> Some id
+  | Path.Pdot _ -> None (* Pdot's can not be unique *)
+  | Path.Papply _ -> assert false
+
+let mark_if_ident p exp uenv =
+  match ident_option_from_path p with
+  | Some id -> mark_seen id (Seen_as id) exp uenv
+  | None -> uenv
+
+let is_unique_variable_opt exp_opt =
+  match exp_opt with
+  | Some ({ exp_desc = Texp_ident(p, _, _, _) } as exp') ->
+      begin match ident_option_from_path p with
+      | Some id -> Some (id, has_unique_attr_texp exp')
+      | None -> None
+      end
+  | _ -> None
+
+let rec exp_to_parent exp =
```

Comment:
I agree that the logic is similar and have moved them closer together. But I think they can not be unified into one function, because they need to be called separately in `check_uniqueness_value_bindings_`.